pkg/controllers/managedzone/suite_test.go

```diff
@@ -0,0 +1,443 @@
+// http://onsi.github.io/ginkgo/ to learn more about Ginkgo.
+
+var (
+	cfg       *rest.Config
+	k8sClient client.Client
+	testEnv   *envtest.Environment
+	ctx       context.Context
+	cancel    context.CancelFunc
+)
+
+const (
+	EventuallyTimeoutMedium   = time.Second * 10
+	ConsistentlyTimeoutMedium = time.Second * 60
+	TestRetryIntervalMedium   = time.Millisecond * 250
+)
+
+func TestAPIs(t *testing.T) {
+	RegisterFailHandler(Fail)
+
+	RunSpecs(t, "Controller Suite")
+}
+
+var _ = BeforeSuite(func() {
+	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
+	ctx, cancel = context.WithCancel(context.TODO())
+
+	By("bootstrapping test environment")
+	testEnv = &envtest.Environment{
+		CRDDirectoryPaths: []string{
+			filepath.Join("../../../", "config", "crd", "bases"),
+			filepath.Join("../../../", "config", "gateway-api", "crd", "standard"),
+			filepath.Join("../../../", "config", "cert-manager", "crd", "v1.7.1"),
+			filepath.Join("../../../", "config", "ocm", "crd"),
+		},
+		ErrorIfCRDPathMissing: true,
+	}
+
+	var err error
+	// cfg is defined in this file globally.
+	cfg, err = testEnv.Start()
+	Expect(err).NotTo(HaveOccurred())
+	Expect(cfg).NotTo(BeNil())
+
+	err = v1alpha1.AddToScheme(scheme.Scheme)
+	Expect(err).NotTo(HaveOccurred())
+
+	err = gatewayv1beta1.AddToScheme(scheme.Scheme)
+	Expect(err).NotTo(HaveOccurred())
+
+	err = certman.AddToScheme(scheme.Scheme)
+	Expect(err).NotTo(HaveOccurred())
+
+	err = workv1.AddToScheme(scheme.Scheme)
+	Expect(err).NotTo(HaveOccurred())
+
+	err = clusterv1beta2.AddToScheme(scheme.Scheme)
+	Expect(err).NotTo(HaveOccurred())
+	//+kubebuilder:scaffold:scheme
+
+	k8sClient, err = client.New(cfg, client.Options{Scheme: scheme.Scheme})
+	Expect(err).NotTo(HaveOccurred())
+	Expect(k8sClient).NotTo(BeNil())
+
+	k8sManager, err := ctrl.NewManager(cfg, ctrl.Options{
+		Scheme:                 scheme.Scheme,
+		HealthProbeBindAddress: "0",
+		MetricsBindAddress:     "0",
+	})
+	Expect(err).ToNot(HaveOccurred())
+
+	dnsProvider := &dns.FakeProvider{}
+
+	err = (&ManagedZoneReconciler{
+		Client:      k8sManager.GetClient(),
+		Scheme:      k8sManager.GetScheme(),
+		DNSProvider: dnsProvider,
+	}).SetupWithManager(k8sManager)
+	Expect(err).ToNot(HaveOccurred())
+
+	err = (&gateway.GatewayClassReconciler{
+		Client: k8sManager.GetClient(),
+		Scheme: k8sManager.GetScheme(),
+	}).SetupWithManager(k8sManager)
+	Expect(err).ToNot(HaveOccurred())
+
+	certificates := tls.NewService(k8sManager.GetClient(), "glbc-ca")
+	dns := dns.NewService(k8sManager.GetClient(), dns.NewSafeHostResolver(dns.NewDefaultHostResolver()), dnsProvider)
+	plc := placement.NewOCMPlacer(k8sManager.GetClient())
+	err = (&gateway.GatewayReconciler{
+		Client:       k8sManager.GetClient(),
+		Scheme:       k8sManager.GetScheme(),
+		Certificates: certificates,
+		Host:         dns,
+		Placement:    plc,
+	}).SetupWithManager(k8sManager, ctx)
+	Expect(err).ToNot(HaveOccurred())
```

Comment:
Do we even need the Gateway and Gateway Class controllers? ManagedZones are essentially a completely separate API with no dependency on gateways.

I do think an integration test should start them all together, but with these pkg level suite_tests i've been conflicted, but that's probably because i don;t think we should have them.

Reply:
if I didn't start this, I couldn't prove that gateways in different namespaces or different hosts didn't get programmed, the status was permanently empty.

Reply:
as you say, we should aim to put these all in one package eventually which would resolve this issue then, when all controllers are running.

Reply:
We solved this kind of problem in the gateway controller integration test with mocking.
e.g. to get the gateway controller reconciler to moved past a certain point, cert manager had to be running and do it's thing. So instead of starting cert manager (since we're not testing that piece) we created a secret that looks like what cert-manager would create.

In this case, you should update the Gateway status directly to get the managedzone controller to continue past whatever point it gets stuck

Reply:
> if I didn't start this, I couldn't prove that gateways in different namespaces or different hosts didn't get programmed, as the gateway's status was permanently empty.

True, but that test would make more sense to me in the gateway_controller test suite since it's that controller doing that

Reply:
This is interesting. I don't think mocking would resolve as I am actually testing what happened in the gateway reconciler as a result of the managed zone, it's not so much that it got stuck. 

Maybe Mike is right here, and this should be moved to the gateway controller test suite... Though I do then wonder if the managed zone controller might need to be started in that test.

I suppose all of these questions are really answered by our plan to amalgamate all of these tests into one package.

Reply:
I will move these tests to the gateway test suite for now.

Reply:
done this now. It does make a lot more sense this way.